### PR TITLE
Egglog RVSDG Roundtrip

### PR DESCRIPTION
**Pull request by oflatt**
Merge after #44

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,7 +9,7 @@ name = "files"
 
 
 [dependencies]
-egglog = { git = "https://github.com/egraphs-good/egglog", rev = "cec39af" }
+egglog = { git = "https://github.com/hatoo/egglog/", rev = "c83a509" }
```
Comment (oflatt) [inline Cargo.toml:12]: Don't merge until this makes it to egglog!

File: src/peg/mod.rs [hunk 2]
```diff
@@ -23,7 +20,7 @@ mod tests;
 #[derive(Debug, PartialEq)]
 pub(crate) enum PegBody {
     /// A basic expression.
-    BasicOp(Expr<Id>),
+    BasicOp(RvsdgExpr<Id>),
```
Comment (ezrosent) [inline src/peg/mod.rs:23]: We're still using this for PEGs, right? should we name this something else?

Also, to Anjali's point earlier, this sort of rename is really easy to review if it's the entirety of the PR (w/the remaining changes stacked on top).

File: src/rvsdg/from_egglog.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            }
+        } else {
+            panic!("expect an operand, got {op}")
+        }
+    }
+
+    fn expr_to_vec_operand(vec: &Expr, bodies: &mut Vec<RvsdgBody>) -> Vec<Operand> {
+        let Expr::Call(func, args) = vec else {
+        panic!("Expected a VO, got {vec}")
+      };
```
Comment (ezrosent) [inline src/rvsdg/from_egglog.rs:28]: formatting looks off here.

File: src/rvsdg/from_egglog.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            }
+        } else {
+            panic!("expect a function, got {expr}")
+        }
+    }
+}
+
+fn vec_map<T>(inputs: &Expr, mut f: impl FnMut(&Expr) -> T) -> Vec<T> {
+    use Expr::*;
+    let mut inputs: &Expr = inputs;
```
Comment (ezrosent) [inline src/rvsdg/from_egglog.rs:206]: declare the inputs parameter as `mut` instead?

File: src/rvsdg/from_egglog.rs [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+                }
+                _ => panic!("expect a function, got {expr}"),
+            }
+        } else {
+            panic!("expect a function, got {expr}")
+        }
+    }
+}
+
+fn vec_map<T>(inputs: &Expr, mut f: impl FnMut(&Expr) -> T) -> Vec<T> {
```
Comment (ezrosent) [inline src/rvsdg/from_egglog.rs:204]: I think this function should be documented. I don't think it's obvious from the signature the way in which it is a map operation.  I'm not positive about what it does, My guess is:

> Call `f` on each element of `inputs`, where `inputs` must be a fully expanded vector egglog vector expression.

File: src/rvsdg/from_egglog.rs [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                        Some(Self::egglog_expr_to_ty(ty)),
+                    )
+                }
+                ("Const", [ty, _const_op, lit]) => RvsdgExpr::Const(
+                    // todo remove the const op from the encoding because it is always ConstOps::Const
+                    ConstOps::Const,
+                    Self::egglog_expr_to_literal(lit),
+                    Self::egglog_expr_to_ty(ty),
+                ),
+                ("PRINT", [opr1, opr2]) => {
```
Comment (ezrosent) [inline src/rvsdg/from_egglog.rs:99]: I think bril supports more than one arg to print (state edge or no), see https://github.com/oflatt/eggcc/blob/main/tests/brils/failing/char/char.bril

File: src/rvsdg/live_variables.rs [hunk 7]
```diff
@@ -173,6 +173,7 @@ impl VarSet {
         self.vars.count_ones(..)
     }
 
+    #[allow(dead_code)]
```
Comment (ezrosent) [inline src/rvsdg/live_variables.rs:176]: If these are dead, can we remove them? (if they're just used in tests, let's annotate them as `#[cfg(test)]`

File: src/rvsdg/schema.egg [hunk 8]
```diff
@@ -41,6 +41,7 @@
 (function not (Type Operand Operand) Expr)
 (function and (Type Operand Operand) Expr)
 (function or (Type Operand Operand) Expr)
+(function PRINT (Operand Operand) Expr)
```
Comment (ezrosent) [inline src/rvsdg/schema.egg:44]: ah, so print isn't taking a vec here?

File: tests/snapshots/files__fib_shape-rvsdg-optimize.snap [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+  v1: int = const 0;
+  v2: int = const 1;
+  v3: int = id v1;
+  v4: int = id v2;
+  v5: int = id v0;
+  jmp .__7__;
+.__130__:
+  v128: int = const 1;
+  v129: bool = eq v121 v128;
+  v68: int = id v87;
```
Comment (ezrosent) [inline tests/snapshots/files__fib_shape-rvsdg-optimize.snap:15]: (no change needed): this is what they mean by "copy insertion" in the "Perfect Reconstruction" paper. 

Sounds like there are standard algorithms here to clean this up.

**Review by ezrosent — CHANGES_REQUESTED** [review]
Seems like this is mostly renames / moves?

File: tests/snapshots/files__fib_shape-rvsdg-optimize.snap (hunk 9 above)
Comment (oflatt) [reply tests/snapshots/files__fib_shape-rvsdg-optimize.snap:15]: ohhhh makes more sense to me now

**Comment by oflatt** [comment]
Yep, this is mostly a cleanup PR

File: Cargo.toml (hunk 1 above)
Comment (oflatt) [reply Cargo.toml]: fixed

File: src/peg/mod.rs (hunk 2 above)
Comment (oflatt) [reply src/peg/mod.rs:23]: Oh I didn't realize it was used in PEGs.
Yeah, renaming should be separate I agree.
I'll do another PR for that!

File: src/rvsdg/from_egglog.rs (hunk 5 above)
Comment (oflatt) [reply src/rvsdg/from_egglog.rs]: good idea, that's right

File: src/rvsdg/from_egglog.rs (hunk 6 above)
Comment (oflatt) [reply src/rvsdg/from_egglog.rs:99]: yeah we just support one for now, we can fix in the future

File: src/rvsdg/live_variables.rs (hunk 7 above)
Comment (oflatt) [reply src/rvsdg/live_variables.rs]: yep, done

**Comment by oflatt** [comment]
Addressed your comments @ezrosent, another rename PR incoming

**Review by ezrosent — APPROVED** [review]

**Pull request merged**